### PR TITLE
feat: Allow reusing same shared IpcSharedMem for transfers

### PR DESCRIPTION
**Pull request by sagudev**
Simple "solution" to https://github.com/servo/ipc-channel/issues/126, it's not exactly zero-copy, but it allows to modify inner buffer of IpcSharedMem and then send it back. This is all `unsafe` but we can uphold all safety variants for `GPUBuffer` (request-response msgs).

Example

before:
```rust
const SIZE: usize = 50;
let (tx1, rx1) = ipc::channel().unwrap();
let (tx2, rx2) = ipc::channel().unwrap();
std::thread::spawn(move || {
    let mut data = rx1.recv().unwrap() // allocation 1
                                      .to_vec(); // allocation 2
    for i in 0..=SIZE {
        *data[i] = i as u8;
    }
    tx2.send(IpcSharedMemory::from_bytes(&data)).unwrap(); // allocation 3
});
tx1.send(IpcSharedMemory::from_byte(0, SIZE)).unwrap();
rx2.recv().unwrap();
```
after:
```rust
const SIZE: usize = 50;
let (tx1, rx1) = ipc::channel().unwrap();
let (tx2, rx2) = ipc::channel().unwrap();
std::thread::spawn(move || {
    // SAFETY: This is safe because IpcSharedMemory is constructed in place from main thread (will only be accessed by send)
    let mut ism = unsafe{ rx1.recv().unwrap() }; // allocation 1
    {
        let data = ism.deref_mut();
        for i in 0..=SIZE {
            *data[i] = i as u8;
        }
    }
    tx2.send(ism).unwrap();
});
tx1.send(IpcSharedMemory::from_byte(0, SIZE)).unwrap();
rx2.recv().unwrap();
```

### CODE REVIEW

**Comment by sagudev** [comment]
Benches from CI: https://github.com/servo/ipc-channel/actions/runs/10472511882

File: src/ipc.rs [hunk 1]
```diff
@@ -570,6 +570,24 @@ impl Deref for IpcSharedMemory {
...
+impl IpcSharedMemory {
+    /// Returns a mutable reference to the deref of this [`IpcSharedMemory`].
+    ///
+    /// # Safety
+    ///
+    /// This is safe if there is only one reader/writer on the data.
+    /// User can achieve this by not cloning [`IpcSharedMemory`]
+    /// and serializing/deserializing only once.
+    #[inline]
+    pub unsafe fn deref_mut(&mut self) -> &mut [u8] {
```
Comment (jschwe) [inline src/ipc.rs:582]: I'm a bit confused on how this works. Do we know that if this code is running, then the other process is guaranteed to not be be able to access this memory?

File: src/platform/unix/mod.rs [hunk 2]
```diff
@@ -866,6 +866,13 @@ impl Deref for OsIpcSharedMemory {
     }
 }
 
+impl OsIpcSharedMemory {
+    #[inline]
+    pub unsafe fn deref_mut(&mut self) -> &mut [u8] {
```
Comment (jschwe) [inline src/platform/unix/mod.rs:871]: In general is `implementing`  deref / deref_mut sound for shared memory? Since Rust / LLVM doesn't see the other process accessing the memory, wouldn't we be required to use volatile reads and writes to prevent unwanted optimizations?

File: src/ipc.rs (hunk 1 above)
Comment (sagudev) [reply src/ipc.rs:582]: no there is no such guarantee, hence the method is marked unsafe (we will probably need named locks and serialization that takes ownership)

File: src/platform/unix/mod.rs (hunk 2 above)
Comment (sagudev) [reply src/platform/unix/mod.rs:871]: no, hence marked as unsafe, same safety docs also applies to this.

File: src/ipc.rs (hunk 1 above)
Comment (sagudev) [reply src/ipc.rs:582]: users needs to make sure that only one process/thread uses memory (no synchronization is guaranteed as user must send ipcsharedmem back), this works if this shared memory is used for something like req-resp, for example webgpu's mapasync where we map (clone) buffer to JS thread (while it's mapped GPU will not touch it) then when unmaping we send this data back.

File: src/platform/unix/mod.rs (hunk 2 above)
Comment (jschwe) [reply src/platform/unix/mod.rs:871]: This question is slightly different - let me rephrase it:
Lets assume the user upholds all Safety invariants - Is it possible that LLVM still breaks our code, because it e.g. sees that the written values are never read (in the current process) and thus optimizes out some writes (or reads)

Comment (sagudev) [reply src/platform/unix/mod.rs:871]: I think not, because they are still read when sending back.

File: src/ipc.rs (hunk 1 above)
Comment (sagudev) [reply src/ipc.rs:582]: I update PR description with example of usage (there is also bench in PR that can act as example).

**Review by jdm — APPROVED** [review]
This seems reasonable as a fully-unsafe API. It would be nice if we could detect invalid usage at runtime similar to Arc::get_mut, but I don't have ideas for how to do that soundly.

**Comment by sagudev** [comment]
> This seems reasonable as a fully-unsafe API. It would be nice if we could detect invalid usage at runtime similar to Arc::get_mut, but I don't have ideas for how to do that soundly.

The main problem is we would need to do cross-process locking which is costly, but we might want to implement this in the future to also fulfill other usecases.
